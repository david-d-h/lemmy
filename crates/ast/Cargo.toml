[package]
name = "ast"
version = "0.1.0"
edition = "2021"

[dependencies]
num-bigint.workspace = true
ecow.workspace = true
typed = { path = "../typed" }
