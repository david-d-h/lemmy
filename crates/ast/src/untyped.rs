use std::rc::Rc;

use ecow::EcoString;

use num_bigint::BigInt;

use crate::{Arg, Statement};

#[derive(Debug, Clone, Eq, PartialEq)]
pub enum Expression {
    String(EcoString),
    Int(BigInt),
    Boolean(bool),
    Identifier(EcoString),
    Group(Rc<Expression>),
    Binary(Rc<Expression>, BinOp, Rc<Expression>),
    Prefix(PrefixOp, Rc<Expression>),
    Postfix(Rc<Expression>, PostfixOp),
    Invocation {
        expr: Rc<Expression>,
        arguments: Vec<Expression>,
    },
    Block(Rc<Vec<Statement>>),
    Function {
        arguments: Vec<Arg>,
        body: Rc<Expression>,
    },
}

pub trait HasBindingPower {
    type Lhs;
    type Rhs;

    fn binding_power(&self) -> (Self::Lhs, Self::Rhs);
}

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub enum BinOp {
    Plus,
    PlusEq,
    Minus,
    MinusEq,
    Star,
    StarEq,
    Slash,
    SlashEq,
    EqEq,
    Neq,
    Gt,
    GtEq,
    Shr,
    ShrEq,
    Lt,
    LtEq,
    Shl,
    ShlEq,
    Amp,
    AmpEq,
    And,
    Caret,
    CaretEq,
    Pipe,
    PipeEq,
    Or,
    Dot,
}

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub enum PrefixOp {
    Minus,
    Plus,
    Bang,
}

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub enum PostfixOp {
    // ...
}

impl HasBindingPower for BinOp {
    type Lhs = u8;
    type Rhs = u8;

    #[inline]
    fn binding_power(&self) -> (u8, u8) {
        match self {
            BinOp::Dot => (24, 23),
            BinOp::Star | BinOp::Slash => (19, 20),
            BinOp::Plus | BinOp::Minus => (17, 18),
            BinOp::Shl | BinOp::Shr => (15, 16),
            BinOp::Amp => (13, 14),
            BinOp::Caret => (11, 12),
            BinOp::Pipe => (9, 10),
            BinOp::EqEq | BinOp::Neq | BinOp::Lt | BinOp::Gt | BinOp::LtEq | BinOp::GtEq => (7, 8),
            BinOp::And => (5, 6),
            BinOp::Or => (3, 4),
            BinOp::PlusEq
            | BinOp::MinusEq
            | BinOp::StarEq
            | BinOp::SlashEq
            | BinOp::ShlEq
            | BinOp::ShrEq
            | BinOp::AmpEq
            | BinOp::CaretEq
            | BinOp::PipeEq => (1, 2),
        }
    }
}

impl HasBindingPower for PrefixOp {
    type Lhs = ();
    type Rhs = u8;

    #[inline]
    fn binding_power(&self) -> ((), u8) {
        match self {
            PrefixOp::Plus | PrefixOp::Minus | PrefixOp::Bang => ((), 21),
        }
    }
}

impl HasBindingPower for PostfixOp {
    type Lhs = u8;
    type Rhs = ();

    #[inline]
    fn binding_power(&self) -> (u8, ()) {
        match self {
            _ => unreachable!(),
        }
    }
}
