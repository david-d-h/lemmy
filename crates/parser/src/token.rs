use core::fmt;

use ecow::EcoString;
use num_bigint::BigInt;

#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub enum TokenKind {
    KeywordLet,
    KeywordFn,
    Identifier,
    Plus,
    PlusEq,
    Minus,
    MinusEq,
    Star,
    StarEq,
    Slash,
    SlashEq,
    Eq,
    EqEq,
    Bang,
    BangEq,
    Gt,
    GtEq,
    Shr,
    ShrEq,
    Lt,
    LtEq,
    Shl,
    ShlEq,
    Amp,
    AmpEq,
    AmpAmp,
    Caret,
    CaretEq,
    Pipe,
    PipeEq,
    PipePipe,
    BigInt,
    Float,
    LitString,
    LitBool,
    Comma,
    Dot,
    LParen,
    RParen,
    LBrace,
    RBrace,
    RArrow,
    Semicolon,
    Colon,
    Eof,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum Token {
    KeywordLet,
    KeywordFn,
    Identifier(EcoString),
    Plus,
    PlusEq,
    Minus,
    MinusEq,
    Star,
    StarEq,
    Slash,
    SlashEq,
    Eq,
    EqEq,
    Bang,
    BangEq,
    Gt,
    GtEq,
    Shr,
    ShrEq,
    Lt,
    LtEq,
    Shl,
    ShlEq,
    Amp,
    AmpEq,
    AmpAmp,
    Caret,
    CaretEq,
    Pipe,
    PipeEq,
    PipePipe,
    BigInt(BigInt),
    LitString(EcoString),
    LitBool(bool),
    Comma,
    Dot,
    LParen,
    RParen,
    LBrace,
    RBrace,
    RArrow,
    Semicolon,
    Colon,
    Eof,
}

impl Token {
    #[inline]
    pub fn kind(&self) -> TokenKind {
        token_to_kind(self)
    }
}

pub fn token_to_kind(token: &Token) -> TokenKind {
    match token {
        Token::KeywordLet => TokenKind::KeywordLet,
        Token::KeywordFn => TokenKind::KeywordFn,
        Token::Identifier(..) => TokenKind::Identifier,
        Token::Plus => TokenKind::Plus,
        Token::PlusEq => TokenKind::PlusEq,
        Token::Minus => TokenKind::Minus,
        Token::MinusEq => TokenKind::MinusEq,
        Token::Star => TokenKind::Star,
        Token::StarEq => TokenKind::StarEq,
        Token::Slash => TokenKind::Slash,
        Token::SlashEq => TokenKind::SlashEq,
        Token::Eq => TokenKind::Eq,
        Token::EqEq => TokenKind::EqEq,
        Token::Bang => TokenKind::Bang,
        Token::BangEq => TokenKind::BangEq,
        Token::Gt => TokenKind::Gt,
        Token::GtEq => TokenKind::GtEq,
        Token::Shr => TokenKind::Shr,
        Token::ShrEq => TokenKind::ShrEq,
        Token::Lt => TokenKind::Lt,
        Token::LtEq => TokenKind::LtEq,
        Token::Shl => TokenKind::Shl,
        Token::ShlEq => TokenKind::ShlEq,
        Token::Amp => TokenKind::Amp,
        Token::AmpEq => TokenKind::AmpEq,
        Token::AmpAmp => TokenKind::AmpAmp,
        Token::Caret => TokenKind::Caret,
        Token::CaretEq => TokenKind::CaretEq,
        Token::Pipe => TokenKind::Pipe,
        Token::PipeEq => TokenKind::PipeEq,
        Token::PipePipe => TokenKind::PipePipe,
        Token::BigInt { .. } => TokenKind::BigInt,
        Token::LitString(_) => TokenKind::LitString,
        Token::LitBool(_) => TokenKind::LitBool,
        Token::Comma => TokenKind::Comma,
        Token::Dot => TokenKind::Dot,
        Token::LParen => TokenKind::LParen,
        Token::RParen => TokenKind::RParen,
        Token::LBrace => TokenKind::LBrace,
        Token::RBrace => TokenKind::RBrace,
        Token::RArrow => TokenKind::RArrow,
        Token::Semicolon => TokenKind::Semicolon,
        Token::Colon => TokenKind::Colon,
        Token::Eof => TokenKind::Eof,
    }
}

impl fmt::Display for TokenKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(match self {
            TokenKind::KeywordLet => "let",
            TokenKind::KeywordFn => "fn",
            TokenKind::Identifier => "<ident>",
            TokenKind::Plus => "+",
            TokenKind::PlusEq => "+=",
            TokenKind::Minus => "-",
            TokenKind::MinusEq => "-=",
            TokenKind::Star => "*",
            TokenKind::StarEq => "*=",
            TokenKind::Slash => "/",
            TokenKind::SlashEq => "/=",
            TokenKind::Eq => "=",
            TokenKind::EqEq => "==",
            TokenKind::Bang => "!",
            TokenKind::BangEq => "!=",
            TokenKind::Gt => ">",
            TokenKind::GtEq => ">=",
            TokenKind::Shr => ">>",
            TokenKind::ShrEq => ">>=",
            TokenKind::Lt => "<",
            TokenKind::LtEq => "<=",
            TokenKind::Shl => "<<",
            TokenKind::ShlEq => "<<=",
            TokenKind::Amp => "&",
            TokenKind::AmpEq => "&=",
            TokenKind::AmpAmp => "&&",
            TokenKind::Caret => "^",
            TokenKind::CaretEq => "^=",
            TokenKind::Pipe => "|",
            TokenKind::PipeEq => "|=",
            TokenKind::PipePipe => "||",
            TokenKind::BigInt => "<number>",
            TokenKind::Float => "<float>",
            TokenKind::LitString => "<string>",
            TokenKind::LitBool => "<bool>",
            TokenKind::Comma => ",",
            TokenKind::Dot => ".",
            TokenKind::LParen => "(",
            TokenKind::RParen => ")",
            TokenKind::LBrace => "{",
            TokenKind::RBrace => "}",
            TokenKind::RArrow => "->",
            TokenKind::Semicolon => ";",
            TokenKind::Colon => ":",
            TokenKind::Eof => "<EOF>",
        })
    }
}

impl fmt::Display for Token {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(match self {
            Token::Identifier(symbol) => symbol.as_str(),
            Token::BigInt(value) => return f.write_str(&value.to_string()),
            Token::LitString(value) => value.as_str(),
            Token::LitBool(value) => {
                if *value {
                    "true"
                } else {
                    "false"
                }
            }
            _ => return token_to_kind(self).fmt(f),
        })
    }
}
