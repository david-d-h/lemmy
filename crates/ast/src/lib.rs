use std::rc::Rc;

use ecow::EcoString;
use untyped::Expression;

pub mod typed;
pub mod untyped;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Statement {
    Expression(Expression),
    Assignment(Assignment),
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Assignment {
    pub value: Rc<Expression>,
    pub name: EcoString,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Arg {
    pub symbol: EcoString,
}
