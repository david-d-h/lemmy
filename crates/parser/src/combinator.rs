use ast::untyped::Expression;

use crate::spanned::{Span, Spanned};
use crate::token::{Token, TokenKind};
use crate::{Parser, Result};

#[inline]
pub fn any_expression(parser: &mut Parser<'_>) -> Result<Expression> {
    crate::expression::expression(parser, None, 0)
}

#[inline]
pub fn just<'src>(kind: TokenKind) -> impl Fn(&mut Parser<'src>) -> Result<Token> {
    move |parser| parser.expect(kind).map(Spanned::into_inner)
}

#[inline]
pub fn map<'src, T, U>(
    inner: impl Fn(&mut Parser<'src>) -> T,
    f: fn(T) -> U,
) -> impl Fn(&mut Parser<'src>) -> U {
    move |parser| f(inner(parser))
}

#[inline]
pub fn try_map<'src, T, U>(
    inner: impl Fn(&mut Parser<'src>) -> Result<T>,
    f: fn(T) -> U,
) -> impl Fn(&mut Parser<'src>) -> Result<U> {
    move |parser| inner(parser).map(f)
}

pub fn with_checkpoint<'src, T>(
    inner: impl Fn(&mut Parser<'src>) -> Result<T>,
) -> impl Fn(&mut Parser<'src>) -> Result<T> {
    move |parser| {
        let backup = parser.lexer.cursor();
        inner(parser).inspect_err(move |_| parser.lexer.insert_cursor_state(backup))
    }
}

#[inline]
pub fn delimited<'src, T>(
    left: TokenKind,
    inner: impl Fn(&mut Parser<'src>) -> Result<T>,
    right: TokenKind,
) -> impl Fn(&mut Parser<'src>) -> Result<T> {
    prefixed([left], terminated(inner, right))
}

pub fn prefixed<'src, const N: usize, T>(
    by: [TokenKind; N],
    inner: impl Fn(&mut Parser<'src>) -> Result<T>,
) -> impl Fn(&mut Parser<'src>) -> Result<T> {
    move |parser| {
        by.try_map(|kind| parser.expect(kind))
            .and_then(|_| inner(parser))
    }
}

pub fn terminated<'src, T>(
    inner: impl Fn(&mut Parser<'src>) -> Result<T>,
    by: TokenKind,
) -> impl Fn(&mut Parser<'src>) -> Result<T> {
    move |parser| inner(parser).and_then(move |out| parser.expect(by).map(|_| out))
}

pub fn maybe<'src, T>(
    inner: impl Fn(&mut Parser<'src>) -> Result<T>,
) -> impl Fn(&mut Parser<'src>) -> Option<T> {
    map(with_checkpoint(inner), Result::ok)
}

pub fn many<'src, T>(
    generator: impl Fn(&mut Parser<'src>) -> Option<T>,
) -> impl Fn(&mut Parser<'src>) -> Result<Vec<T>> {
    move |parser| {
        let mut items = Vec::new();
        while let Some(item) = generator(parser) {
            items.push(item);
        }
        Ok(items)
    }
}

pub fn separated<'src, T>(
    generator: impl Fn(&mut Parser<'src>) -> Result<T>,
    by: TokenKind,
) -> impl Fn(&mut Parser<'src>) -> Result<Vec<T>> {
    move |parser| {
        let mut items = Vec::new();
        while let Some(item) = maybe(&generator)(parser) {
            items.push(item);
            if !parser.eat(by) {
                break;
            }
        }
        if items.len() != 0 {
            _ = maybe(just(by))(parser);
        }
        Ok(items)
    }
}

pub fn spanned<'src, T>(
    inner: impl Fn(&mut Parser<'src>) -> T,
) -> impl Fn(&mut Parser<'src>) -> Spanned<T> {
    move |parser| {
        let start = parser.lexer.cursor().anchor;
        let item = inner(parser);
        let end = parser.lexer.cursor().current;
        Spanned {
            span: Span { start, end },
            item,
        }
    }
}
