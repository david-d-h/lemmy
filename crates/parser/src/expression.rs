use std::rc::Rc;

use ast::untyped as ast;
use ast::*;

use crate::combinator::*;
use crate::token::Token;
use crate::token::TokenKind;
use crate::{Parser, Result};

#[inline]
pub fn bp<T: HasBindingPower>(of: T) -> (T::Lhs, T::Rhs) {
    of.binding_power()
}

/// Parse an expression with operator and operator presedence in mind.
pub fn expression(
    parser: &mut Parser<'_>,
    stack: Option<Expression>,
    power: u8,
) -> Result<Expression> {
    let peek = parser.peek_kind();

    match stack {
        // parse an invocation of a lhs expression
        Some(lhs) if let Some(TokenKind::LParen) = peek => {
            let stack = Some(invocation(parser, lhs)?);
            expression(parser, stack, power)
        },
        Some(lhs) // finish a bound postfix expression and recurse
            if let Some(op) = peek.map(to_postfix_op).flatten()
                && let (lhs_bp, ()) = bp(op)
                && lhs_bp >= power =>
        {
            parser
                .advance()
                .expect("a postfix unary operator was just peeked");
            expression(parser, Some(Expression::Postfix(Rc::new(lhs), op)), power)
        },
        Some(lhs) // parse rhs of binary operator expression, finish, and recurse
            if let Some(op) = peek.map(to_bin_op).flatten()
                && let (lhs_bp, rhs_bp) = bp(op)
                && lhs_bp >= power =>
        {
            parser.advance().expect("a binary operator was just peeked");
            let unconcerned_rhs = unconcerned_expr(parser)?;
            let rhs = expression(parser, Some(unconcerned_rhs), rhs_bp)?;
            expression(parser, Some(Expression::Binary(Rc::new(lhs), op, Rc::new(rhs))), power)
        },
        // a prefix operator was found, parse a rhs expression and recurse
        None if let Some(op) = peek.map(to_prefix_op).flatten() => {
            parser
                .advance()
                .expect("a prefix unary operator was just peeked");
            let ((), rhs_bp) = bp(op);
            let rhs = expression(parser, None, rhs_bp)?;
            expression(parser, Some(Expression::Prefix(op, Rc::new(rhs))), power)
        },
        Some(expr) => Ok(expr), // finished parsing a singular expression
        None => { // start parsing a singular expression that may (or may not) include operators
            let unconcerned = unconcerned_expr(parser)?;
            expression(parser, Some(unconcerned), power)
        },
    }
}

pub fn unconcerned_expr(parser: &mut Parser<'_>) -> Result<Expression> {
    let token = parser.peek_one_of([
        TokenKind::Identifier,
        TokenKind::BigInt,
        TokenKind::LitString,
        TokenKind::LitBool,
        TokenKind::LParen,
        TokenKind::LBrace,
        TokenKind::KeywordFn,
    ])?;

    let expr = match token.into_inner() {
        Token::Identifier(symbol) => Expression::Identifier(symbol),
        Token::BigInt(value) => Expression::Int(value),
        Token::LitString(value) => Expression::String(value),
        Token::LitBool(value) => Expression::Boolean(value),
        Token::LParen => return group(parser),
        Token::LBrace => return block(parser),
        Token::KeywordFn => return function(parser),
        _ => unreachable!(),
    };

    _ = parser
        .advance()
        .expect("peek was successful, so advance should be too?");

    Ok(expr)
}

pub fn group(parser: &mut Parser<'_>) -> Result<Expression> {
    delimited(TokenKind::LParen, any_expression, TokenKind::RParen)(parser)
        .map(Rc::new)
        .map(Expression::Group)
}

pub fn block(parser: &mut Parser<'_>) -> Result<Expression> {
    delimited(
        TokenKind::LBrace,
        many(maybe(crate::statement::statement)),
        TokenKind::RBrace,
    )(parser)
    .map(Rc::new)
    .map(Expression::Block)
}

pub fn function(parser: &mut Parser<'_>) -> Result<Expression> {
    Ok(Expression::Function {
        arguments: prefixed([TokenKind::KeywordFn], crate::common::args)(parser)?,
        body: Rc::new(any_expression(parser)?),
    })
}

pub fn invocation(parser: &mut Parser<'_>, on: Expression) -> Result<Expression> {
    delimited(
        TokenKind::LParen,
        separated(any_expression, TokenKind::Comma),
        TokenKind::RParen,
    )(parser)
    .map(move |arguments| Expression::Invocation {
        expr: Rc::new(on),
        arguments,
    })
}

#[cfg(test)]
mod tests {
    use crate::lexer::Lexer;

    use super::*;

    fn expression(src: &'static str) -> Expression {
        any_expression(&mut Parser {
            lexer: Lexer::new(src),
        })
        .expect("source should consist of a valid expression")
    }

    crate::insta_tests![
        [compact] string as expression { "Hallo!" }

        [compact] number as expression { 1 }

        [compact] boolean as expression { true }

        [compact] ident as expression { hallo_ }

        [compact] grouped_ident as expression { (hal_lo_) }

        nested_groups as expression { (((a))) }

        [compact] one_plus_one as expression { 1 + 1 }

        [compact] accessor as expression { person.name }

        [compact] presedence_mul_over_plus as expression { 1 + 1 * 2 }

        [compact] negative_number_literal as expression { -12 }

        [compact] invocation_of_ident as expression { real(1, 2, 3) }

        [compact] invocation_of_ident_with_trailing_comma_in_argument_list as expression { real(1, 2,) }

        block_many_expr as expression {
            {
                io.print("harharhar")
                io.debug(3, 2, 1)
            }
        }

        [compact] simple_function as expression { fn(x) x }

        simple_function_with_block as expression { fn(x) { x } }
    ];
}

pub const fn to_bin_op(kind: TokenKind) -> Option<BinOp> {
    match kind {
        TokenKind::Plus => Some(BinOp::Plus),
        TokenKind::PlusEq => Some(BinOp::PlusEq),
        TokenKind::Minus => Some(BinOp::Minus),
        TokenKind::MinusEq => Some(BinOp::MinusEq),
        TokenKind::Star => Some(BinOp::Star),
        TokenKind::StarEq => Some(BinOp::StarEq),
        TokenKind::Slash => Some(BinOp::Slash),
        TokenKind::SlashEq => Some(BinOp::SlashEq),
        TokenKind::EqEq => Some(BinOp::EqEq),
        TokenKind::BangEq => Some(BinOp::Neq),
        TokenKind::Gt => Some(BinOp::Gt),
        TokenKind::GtEq => Some(BinOp::GtEq),
        TokenKind::Shr => Some(BinOp::ShrEq),
        TokenKind::Lt => Some(BinOp::Lt),
        TokenKind::LtEq => Some(BinOp::LtEq),
        TokenKind::Shl => Some(BinOp::Shl),
        TokenKind::ShlEq => Some(BinOp::ShlEq),
        TokenKind::Amp => Some(BinOp::Amp),
        TokenKind::AmpEq => Some(BinOp::AmpEq),
        TokenKind::AmpAmp => Some(BinOp::And),
        TokenKind::Caret => Some(BinOp::Caret),
        TokenKind::CaretEq => Some(BinOp::CaretEq),
        TokenKind::Pipe => Some(BinOp::Pipe),
        TokenKind::PipeEq => Some(BinOp::PipeEq),
        TokenKind::PipePipe => Some(BinOp::Or),
        TokenKind::Dot => Some(BinOp::Dot),
        _ => None,
    }
}

pub const fn to_prefix_op(kind: TokenKind) -> Option<PrefixOp> {
    match kind {
        TokenKind::Plus => Some(PrefixOp::Plus),
        TokenKind::Minus => Some(PrefixOp::Minus),
        TokenKind::Bang => Some(PrefixOp::Bang),
        _ => None,
    }
}

pub const fn to_postfix_op(kind: TokenKind) -> Option<PostfixOp> {
    match kind {
        _ => None,
    }
}
