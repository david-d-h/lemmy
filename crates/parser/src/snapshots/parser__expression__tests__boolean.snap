---
source: crates/parser/src/expression.rs
expression: "expression (:: core :: stringify! (true))"
snapshot_kind: text
---
Boolean(true)
